docs/control-structures.rst

```diff
@@ -167,3 +167,397 @@ Currently, there are three situations, where exceptions happen automatically in
 
 Internally, Solidity performs an "invalid jump" when an exception is thrown and thus causes the EVM to revert all changes made to the state. The reason for this is that there is no safe way to continue execution, because an expected effect did not occur. Because we want to retain the atomicity of transactions, the safest thing to do is to revert all changes and make the whole transaction (or at least call) without effect.
 
+.. index:: ! assembly, ! asm, ! evmasm
+
+Inline Assembly
+===============
+
+For more fine-grained control especially in order to enhance the language by writing libraries,
+it is possible to interleave Solidity statements with inline assembly in a language close
+to the one of the virtual machine. Due to the fact that the EVM is a stack machine, it is
+often hard to address the correct stack slot and provide arguments to opcodes at the correct
+point on the stack. Solidity's inline assembly tries to facilitate that and other issues
+arising when writing manual assembly by the following features:
+
+* functional-style opcodes: `mul(1, add(2, 3))` instead of `push1 3 push1 2 add push1 1 mul`
+* assembly-local variables: `let x := add(2, 3)  let y := mload(0x40)  x := add(x, y)`
+* access to external variables: `function f(uint x) { assembly { x := sub(x, 1) } }`
+* labels: `let x := 10  repeat: x := sub(x, 1) jumpi(repeat, eq(x, 0))`
+
+We now want to describe the inline assembly language in detail.
+
+.. warning::
+	Inline assembly is still a relatively new feature and might change if it does not prove useful,
+	so please try to keep up to date.
+
+Example
+-------
+
+The following example provides library code to access the code of another contract and
+load it into a `bytes` variable.
+
+.. code::
+
+	library GetCode {
+		function at(address _addr) returns (bytes o_code) {
+			assembly {
+				let size := extcodesize(_addr);
+				// allocate output byte array - could also just do that using "new"
+				o_code := mload(0x40)
+				// new "memory end" including padding
+				mstore(0x40, add(o_code, and(add(add(size, 0x20), 0x1f), bnot(0x1f))))
+				// store length in memory
+				mstore(o_code, size)
+				// actually retrieve the code
+				extcodecopy(_addr, add(o_code, 0x20), 0, size)
+			}
+		}
+	}
+
```

Comment:
Perhaps something here in the docs that could show a side by side comparison of how you would do it "the normal way" and then show the estimated gas used....something that could really show the benefits of utilizing this.